mlir/include/Quantum/Utils/QuantumSplitting.h

```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+namespace catalyst {
+namespace quantum {
+
+/// A collection of the data required to reconstruct a deterministic hybrid quantum program with
+/// classical preprocessing and arbitrary classical control flow.
+struct QuantumCache {
+    mlir::TypedValue<ArrayListType> paramVector;
+    mlir::TypedValue<ArrayListType> wireVector;
+    /// For every structured control flow op, store the values required for it to execute.
+    /// Specifically: store the conditions for scf.if ops, the start/stop/step of scf.for ops, and
+    /// the number of iterations for scf.while ops.
+    mlir::DenseMap<mlir::Operation *, mlir::TypedValue<ArrayListType>> controlFlowTapes;
+
+    /// Initialize the quantum cache to traverse and store the necessary parameters for the given
+    /// `topLevelRegion`.
+    static QuantumCache initialize(mlir::Region &topLevelRegion, mlir::OpBuilder &builder,
+                                   mlir::Location loc);
+};
+
+class AugmentedCircuitGenerator {
+  public:
+    AugmentedCircuitGenerator(mlir::IRMapping &oldToCloned, mlir::PatternRewriter &rewriter,
+                              QuantumCache &cache)
+        : oldToCloned(oldToCloned), rewriter(rewriter), cache(cache)
+    {
+    }
+
+    /// Given a `region` containing classical preprocessing and quantum operations, generate an
+    /// augmented version that caches all the parameters required to deterministically re-execute
+    /// the circuit (gate params, classical control flow, and dynamic wires).
+    void generate(mlir::Region &region);
+
+  private:
+    mlir::IRMapping &oldToCloned;
+    mlir::PatternRewriter &rewriter;
```

Comment:
Just a comment: I think it would be safer to pass rewriters explicitly via function arguments rather than keeping it as a global variable to alter with insertion guards. Don't see any problems right now though.

Reply:
I vaguely recall making the decision to have the rewriter/builder as a struct property last week but I don't remember why. I'll see if it's a lot of effort to refactor to passing it as a function argument because I agree that it's perhaps more idiomatic.